### PR TITLE
DLIS-6165 Improve input mismatch error for inference requests

### PR DESCRIPTION
**Pull request by indrajit96**
Created 2 new strings to improve the verbosity of the error response.

Scenario: When a inference request has less input params than the model expects error response is insufficient.
Fix: Added 2 new strings to the error response

Old Behavior:
[request id: <id_unknown>] expected 5 inputs but got 1 inputs for model 'llama2-7b'

New Behavior:
[request id: <id_unknown>] expected 5 inputs but got 1 inputs for model 'llama2-7b'. Got inputs ['prompt'], but missing ['max_output_token', 'top_k', 'top_p', 'temperature']"

We have added text to existing response so that existing tests do not break.
Working on the test case for this response will add later.

### CODE REVIEW

File: src/infer_request.cc [hunk 1]
```diff
@@ -1006,12 +1006,50 @@ InferenceRequest::Normalize()
...
+      if (original_inputs_.size() == 0) {
+        original_input_string = "[]";
+      }
       return Status(
           Status::Code::INVALID_ARG,
           LogRequest() + "expected " +
               std::to_string(model_config.input_size()) + " inputs but got " +
               std::to_string(original_inputs_.size()) + " inputs for model '" +
-              ModelName() + "'");
+              ModelName() + "'." + "Got inputs " + original_input_string +
```
Comment (rmccorm4) [inline src/infer_request.cc:1051]:
```suggestion
              ModelName() + "'. Got inputs " + original_input_string +
```

File: src/infer_request.cc [hunk 2]
```diff
@@ -1006,12 +1006,50 @@ InferenceRequest::Normalize()
...
+      // the inputs and then the string this way saves memory and overhead of
+      // creating a vector.
+      original_input_string.pop_back();
+      original_input_string = original_input_string + "]";
+      if (missing_input_cnt == 0) {
+        missing_input_string = "[]";
+      }
+      if (original_inputs_.size() == 0) {
+        original_input_string = "[]";
+      }
```
Comment (rmccorm4) [inline src/infer_request.cc:1045]: We'll want to test a few edge cases here in some new tests:
1. some required + some optional inputs
    - say a model has 2 required inputs and 2 optional inputs, but 1 required input is missing. Will the error message say something like `"Got inputs [required0], but missing [required1, optional0, optional1]"` or will it say `"Got inputs [required0], but missing [required1]"`. We may need some extra logic to distinguish between required and optional in these messages.

2. can `missing_count == 0` case happen? wouldn't this imply we had all the inputs?
3. should test `original_inputs_.size() == 0` just to automate validating we get the right "[]" empty string you added

File: src/infer_request.cc [hunk 3]
```diff
@@ -1006,12 +1006,50 @@ InferenceRequest::Normalize()
...
+      missing_input_string = missing_input_string + "]";
+
+      for (const auto& pair : original_inputs_) {
+        original_input_string =
+            original_input_string + "'" + pair.first + "'" + ",";
+      }
+      // Removes the extra ","
+      // Can be handled in a a more elegant way by creating a vector first of
+      // the inputs and then the string this way saves memory and overhead of
+      // creating a vector.
```
Comment (rmccorm4) [inline src/infer_request.cc:1037]: Don't need to duplicate the comment in both places, and I don't think you need the "can be handled in a more elegant way part". Just saying `Removes the extra ","` once is sufficient I think.

File: src/infer_request.cc (hunk 2 above)
Comment (rmccorm4) [reply src/infer_request.cc:1045]: I can't find a good place to append these tests to, so I think we can just add a new test. Maybe named something like `L0_input_validation`. You can probably use [L0_nan_inf](https://github.com/triton-inference-server/server/tree/main/qa/L0_nan_inf) as a relatively simple reference for how to write a new test.

Basically you need 3 things:
1. `models/` - some example model repository containing simple models used to reproduce your error messages / test cases
2. `input_validation_test.py` - defines the unit tests for various cases, i.e. setting up clients, sending requests to different models, and validating the error messages returned
3. `test.sh` - starts tritonserver using your test model repository, then runs your python unit tests

We can chat more offline about the details of adding/writing tests if you have questions

Comment (indrajit96) [reply src/infer_request.cc]: Taken care in the server repo PR
https://github.com/triton-inference-server/server/pull/6933

Comment (indrajit96) [reply src/infer_request.cc]: Also added extra logic to handle optional inputs

**Comment by lkomali** [comment]
The message can be improved for better clarity and readability to the below message.
[request ID: <id_unknown>] Error: Expected 5 inputs for model 'llama2-7b', but received only 1 input(s).
Received inputs: ['prompt']
Missing inputs: ['max_output_token', 'top_k', 'top_p', 'temperature']
Please provide all required inputs.

**Comment by indrajit96** [comment]
> The message can be improved for better clarity and readability to something similar to the one below.
> 
> [request ID: <id_unknown>] Error: Expected 5 inputs for model 'llama2-7b', but received only 1 input(s). Received inputs: ['prompt'] Missing inputs: ['max_output_token', 'top_k', 'top_p', 'temperature'] Please provide all required inputs.

Fixed this made the message more verbose.

File: src/infer_request.cc [hunk 4]
```diff
@@ -1005,21 +1005,54 @@ InferenceRequest::Normalize()
       (original_inputs_.size() < model_raw_->RequiredInputCount())) {
     // If no input is marked as optional, then use exact match error message
     // for consistency / backward compatibility
+    std::string missing_required_input_string = "[";
```
Comment (rmccorm4) [inline src/infer_request.cc:1008]: This Normalize() function is already quite long. What do you think about refactoring this block into a helper function?
```cpp
InferenceRequest::Normalize() {
  ...
  // Make sure that the request is providing the number of inputs
  // as is expected by the model.
  RETURN_IF_ERROR(ValidateRequestInputs(...));

  ...
}

Status ValidateRequestInputs(...) {
    if ((original_inputs_.size() > (size_t)model_config.input_size()) ||
      (original_inputs_.size() < model_raw_->RequiredInputCount())) {
      ...
        if ((size_t)model_config.input_size() == model_raw_->RequiredInputCount()) { 
          return Status(...)
       } else {
         return Status(...)
       }
    }

    return Status::Success;
}
```

Comment (indrajit96) [reply src/infer_request.cc:1008]: Yes this looks neat.
Will update the PR with a quick round of test

**Comment by kthui** [comment]
Please update the copyright year on both files, because this change is more than fixing a typo, i.e. on infer_request.h, the copyright should start with
```
// Copyright 2020-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
same for infer_request.cc

**Comment by indrajit96** [comment]
> Please update the copyright year on both files, because this change is more than fixing a typo, i.e. on infer_request.h, the copyright should start with
> 
> ```
> // Copyright 2020-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
> ```
> 
> same for infer_request.cc

Fixed for all files

**Comment by indrajit96** [comment]
> Please update the copyright year on both files, because this change is more than fixing a typo, i.e. on infer_request.h, the copyright should start with
> 
> ```
> // Copyright 2020-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
> ```
> 
> same for infer_request.cc

Fixed for all files

**Review by kthui — APPROVED** [review]
LGTM!

**Review by rmccorm4 — APPROVED** [review]
Looks great!
